types/consensus_data.go

```diff
@@ -127,7 +129,7 @@ type ConsensusData struct {
 	// Current 30M gas limit produces 30M / 16 (call data cost) = 1,875,000 bytes (https://eips.ethereum.org/EIPS/eip-4488)
 	// we can upper limit transactions to 2^21, together with the rest of the data 2*2^21 = 2^22 = 4,194,304 bytes
 	// Exmplanation on why transaction sizes are so big https://github.com/ethereum/consensus-specs/pull/2686
-	DataSSZ []byte `ssz-max:"4194304"` // 2^22
+	DataSSZ []byte `ssz-max:"4981024"` // 2^22 + 288 + 786432
```

Comment:
There is a slight discrepancy in the `ssz-max` value for `DataSSZ`. The current `ssz-max` is set to `4,981,024` bytes, but based on a detailed breakdown of the components (including `KZG Proofs`, `Blobs`, basic `BeaconBlock` components, and the fixed-size elements in `BeaconBlockBody`), the calculated maximum size should be `4,981,464` bytes.

This difference of `440 bytes` seems to arise from not fully accounting for:
- The fixed-size components within `BeaconBlockBody` (`RANDAOReveal`, `ETH1Data`, `Graffiti`, `SyncAggregate`), which total `360 bytes`.
- The basic `BeaconBlock` size (components `Slot`, `ProposerIndex`, `ParentRoot`, `StateRoot`), adding up to `80 bytes`.

Additionally, we need to adjust the full-data `ssz-max` in SignedMessage accordingly

Reply:
I've fixed some fields that we missed and I actually arrived at a bigger value: 5305388.
@GalRogozinski We would like a third opinion on that. Could you calculate your ssz-max value for DataSSZ (Deneb.BlockContents)?
If you'll push anything, pls don't forget to generate the ssz and extend the ssz-max values to FullData in SignedMessage

Reply:
Looking into the calculation:
`Attestation(2048*8 + 96 + AttestationData(2*8 + 32 + 2*(8+32)))`

This is derived from:
```
type Attestation struct {
	AggregationBits bitfield.Bitlist `ssz-max:"2048"`
	Data            *AttestationData
	Signature       BLSSignature `ssz-size:"96"`
}
```
So not sure why we multiply 2048 by 8?
I am not even sure why `ssz-max:"2048"` is correct?
There should be max 2048 bits in the `Bitlist` because this is the max number of committee members. But ssz sizes are given in bytes everywhere so it should be 2048/8 = 256...

Probably the go-eth2-client lib is right and I am wrong but I don't understand why...

Reply:
The following python code prints:
`Total Size without ExecutionPayload: 3207032`

That's slightly different than what Matheos calculated. This doesn't take my previous comment into account

```python
# Constants
KZG_PROOFS_SIZE = 6 * 48  # KZGProofs size
BLOBS_SIZE = 6 * 131072  # Blobs size
BEACON_BLOCK_OVERHEAD = 2 * 32 + 2 * 8  # Additional overhead for BeaconBlock

# Components of BeaconBlockBody
ETH1_DATA_SIZE = 96 + 2 * 32 + 8  # ETH1Data
PROPOSER_SLASHING_SIZE = 16 * (2 * (96 + 3 * 32 + 2 * 8))  # ProposerSlashing
ATTESTER_SLASHING_SIZE = 2 * (2 * (2048 * 8 + 96 + (2 * 8 + 32 + 2 * (8 + 32))))  # AttesterSlashing
ATTESTATION_SIZE = 128 * (2048 * 8 + 96 + (2 * 8 + 32 + 2 * (8 + 32)))  # Attestation
DEPOSIT_SIZE = 16 * (33 * 32 + 48 + 32 + 8 + 96)  # Deposit
SIGNED_VOLUNTARY_EXIT_SIZE = 16 * (96 + 2 * 8)  # SignedVoluntaryExit
SYNC_AGGREGATE_SIZE = 64 + 96  # SyncAggregate
BLS_TO_EXECUTION_CHANGES_SIZE = 16 * (96 + (8 + 48 + 20))  # BLSToExecutionChanges
KZG_COMMITMENT_SIZE = 4096 * 48  # KZGCommitment

# BeaconBlockBody total size without ExecutionPayload
beacon_block_body_size_without_execution_payload = (
    ETH1_DATA_SIZE + PROPOSER_SLASHING_SIZE + ATTESTER_SLASHING_SIZE + 
    ATTESTATION_SIZE + DEPOSIT_SIZE + SIGNED_VOLUNTARY_EXIT_SIZE + 
    SYNC_AGGREGATE_SIZE + BLS_TO_EXECUTION_CHANGES_SIZE + KZG_COMMITMENT_SIZE
)

# Total size of Deneb.BlockContents and BeaconBlock without ExecutionPayload
total_size_without_execution_payload = KZG_PROOFS_SIZE + BLOBS_SIZE + BEACON_BLOCK_OVERHEAD + beacon_block_body_size_without_execution_payload

# Output the total size
print("Total Size without ExecutionPayload:", total_size_without_execution_payload)
```

Reply:
When I take the comment into account I get 1,078,136

```python
# Constants
KZG_PROOFS_SIZE = 6 * 48  # KZGProofs size
BLOBS_SIZE = 6 * 131072  # Blobs size
BEACON_BLOCK_OVERHEAD = 2 * 32 + 2 * 8  # Additional overhead for BeaconBlock

# Components of BeaconBlockBody
ETH1_DATA_SIZE = 96 + 2 * 32 + 8  # ETH1Data
PROPOSER_SLASHING_SIZE = 16 * (2 * (96 + 3 * 32 + 2 * 8))  # ProposerSlashing
ATTESTER_SLASHING_SIZE = 2 * (2 * (2048 / 8 + 96 + (2 * 8 + 32 + 2 * (8 + 32))))  # AttesterSlashing
ATTESTATION_SIZE = 128 * (2048/8 + 96 + (2 * 8 + 32 + 2 * (8 + 32)))  # Attestation
DEPOSIT_SIZE = 16 * (33 * 32 + 48 + 32 + 8 + 96)  # Deposit
SIGNED_VOLUNTARY_EXIT_SIZE = 16 * (96 + 2 * 8)  # SignedVoluntaryExit
SYNC_AGGREGATE_SIZE = 64 + 96  # SyncAggregate
BLS_TO_EXECUTION_CHANGES_SIZE = 16 * (96 + (8 + 48 + 20))  # BLSToExecutionChanges
KZG_COMMITMENT_SIZE = 4096 * 48  # KZGCommitment

# BeaconBlockBody total size without ExecutionPayload
beacon_block_body_size_without_execution_payload = (
    ETH1_DATA_SIZE + PROPOSER_SLASHING_SIZE + ATTESTER_SLASHING_SIZE + 
    ATTESTATION_SIZE + DEPOSIT_SIZE + SIGNED_VOLUNTARY_EXIT_SIZE + 
    SYNC_AGGREGATE_SIZE + BLS_TO_EXECUTION_CHANGES_SIZE + KZG_COMMITMENT_SIZE
)

# Total size of Deneb.BlockContents and BeaconBlock without ExecutionPayload
total_size_without_execution_payload = KZG_PROOFS_SIZE + BLOBS_SIZE + BEACON_BLOCK_OVERHEAD + beacon_block_body_size_without_execution_payload

# Output the total size
print("Total Size without ExecutionPayload:", total_size_without_execution_payload)
```

Reply:
Also notice 2^21 is actually above the estimation so not sure how strict we should be

Reply:
`ssz-max:"2048"` is also strange for me. But, since it's in go-eth2-client, I believe we should, for now, take that into account in the calculation.
On the other hand, 2048 * 8 seems to be a mistake for me. I think we should use just 2048. I.e.:
```python
ATTESTER_SLASHING_SIZE = 2 * (2 * (2048 + 96 + (2 * 8 + 32 + 2 * (8 + 32))))  # AttesterSlashing
ATTESTATION_SIZE = 128 * (2048 + 96 + (2 * 8 + 32 + 2 * (8 + 32)))  # Attestation
```
I think in the script, it's missing the graffiti:
```python
ETH1_DATA_SIZE = 96 + (2 * 32 + 8) + 32 # Randao + eth1data + graffiti
```
And if we consider the `ExecutionPayload` without the transactions, we would have:
```python
EXECUTION_PAYLOAD_NO_TRANSACTIONS = 32 + 20 + 2*32 + 256 + 32 + 4*8 + 3*32 + 16 * (2*8 + 20 + 8) + 8 + 8
```
Running it, I get 1,315,964 for everything but transactions. Adding 2^21 as the transactions upper limit, I get 3,413,116.
What do you think @GalRogozinski ?

Reply:
@MatheusFranco99 @GalRogozinski 2048 is bits I believe, not bytes.

```
64 attester committees * 2048 members per committee * 32 slots per epoch * 32 eth = 134,217,728
```

There's not enough ETH supply to have 2048 members per attester committee, so I'd guess the size is `2048 / 8` (bits to bytes).

Reply:
Additionally, see the Beacon spec increasing GOSSIP_MAX_SIZE from 1mb to 10mb post-bellatrix: https://github.com/ethereum/consensus-specs/blob/3727a757450feb6110e2d4e403604572a478343d/specs/bellatrix/p2p-interface.md?plain=1#L147-L153

Reply:
https://github.com/ethereum/consensus-specs/blob/3727a757450feb6110e2d4e403604572a478343d/specs/bellatrix/p2p-interface.md?plain=1#L155-L161

```
Considering that `BeaconBlock` max size is on the
order of 128 KiB in the worst case and the current gas limit (~30M) bounds max blocksize to less
than 2 MiB today, this marginal difference in theoretical bounds will have zero
impact on network functionality and security.
```

Reply:
Yesterday I opened https://github.com/attestantio/go-eth2-client/issues/98 because of the 2048 bytes issue.
For the calculation we carried over the issue because we rely on the lib... but if all 3 of us agree it is a mistake on go-eth2-client side we can fix the calculation already.

Besides this I think the current calculation is more precise than yours because we go into details about the block structure.

About the 10% margin. According to the spec: 
`the current gas limit (~30M) bounds max blocksize to less
than 2 MiB today`
So we already have some margin by assuming 2MiB.

About 40M gas, for eth to change that I think it must be in a hard-fork. So I don't see a need to prepare in advanced... am I missing something?

Reply:
Personally, though I think it's correct, I don't feel comfortable using 2048/8 while their ssz-max is 2048 bytes.

About the transactions and gas, I could go either way, as you guys prefer. I think it's both easy to keep 30M and to switch to 40.